File: rainbow-lib/src/main/java/com/juankysoriano/rainbow/core/event/RainbowInputController.java
package com.juankysoriano.rainbow.core.event;

import android.support.annotation.NonNull;
import android.view.MotionEvent;

import com.juankysoriano.rainbow.core.drawing.RainbowDrawer;

public class RainbowInputController {
    private static final int DIVISIONS = 3;
    private final RainbowDrawer rainbowDrawer;
    private final FingerPositionSmoother smoother;
    private float x, y;
    private float px, py;
    private RainbowInteractionListener rainbowInteractionListener;
    private boolean screenTouched;
    private boolean fingerMoving;
    private float scaleFactor;
    private MotionEvent nextEvent;

    public static RainbowInputController newInstance() {
        RainbowDrawer rainbowDrawer = new RainbowDrawer();
        FingerPositionSmoother positionSmoother = new FingerPositionSmoother();
        return new RainbowInputController(rainbowDrawer, positionSmoother);
    }

    private RainbowInputController(RainbowDrawer rainbowDrawer,
                                   FingerPositionSmoother predictor) {
        this.rainbowDrawer = rainbowDrawer;
        smoother = predictor;
        x = y = px = py = -1;
    }

    public void setScale(float scaleFactor) {
        this.scaleFactor = scaleFactor;
    }

    public void postEvent(final MotionEvent motionEvent) {
        nextEvent = motionEvent;
    }

    public void dispatchEvent() {
        if (nextEvent == null) {
            return;
        }

        switch (nextEvent.getAction()) {
            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_DOWN:
                process(nextEvent);
                break;
            case MotionEvent.ACTION_MOVE:
                splitIntoMultipleEvents(nextEvent);
                break;
            default://no-op
        }
        nextEvent = null;
    }

    public void clearEvents() {
        nextEvent = null;
    }

    private void process(@NonNull MotionEvent motionEvent) {
        MotionEvent scaledEvent = MotionEvent.obtain(motionEvent);
        scaledEvent.setLocation(motionEvent.getX() * scaleFactor, motionEvent.getY() * scaleFactor);
        preHandleEvent(scaledEvent);
        handleSketchEvent(scaledEvent);
        postHandleEvent();
    }

    private void splitIntoMultipleEvents(@NonNull MotionEvent event) {
        float px = getX();
        float py = getY();
        float diffX = event.getX() - px;
        float diffY = event.getY() - py;

        for (int i = 1; i <= DIVISIONS; i++) {
            float newEventX = px + diffX * i / DIVISIONS;
            float newEventY = py + diffY * i / DIVISIONS;
            MotionEvent subEvent = obtainEventWithNewPosition(event, newEventX, newEventY);
            process(subEvent);
        }
    }

    private MotionEvent obtainEventWithNewPosition(@NonNull MotionEvent event, float newEventX, float newEventY) {
        MotionEvent motionEvent = MotionEvent.obtain(event);
        motionEvent.setLocation(newEventX, newEventY);
        return motionEvent;
    }

    private void preHandleEvent(MotionEvent motionEvent) {
        x = motionEvent.getX();
        y = motionEvent.getY();
    }

    private void handleSketchEvent(final MotionEvent event) {

        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                screenTouched = true;
                fingerMoving = false;
                performTouch(event);
                break;
            case MotionEvent.ACTION_UP:
                screenTouched = false;
                fingerMoving = false;
                performRelease(event);
                break;
            case MotionEvent.ACTION_MOVE:
                screenTouched = true;
                fingerMoving = true;
                performMove(event);
                break;
        }

        performMotion(event);
    }

    private void postHandleEvent() {
        px = x;
        py = y;
    }

    private void performTouch(MotionEvent event) {
        smoother.resetTo(event.getX(), event.getY());
        if (rainbowInteractionListener != null) {
            rainbowInteractionListener.onSketchTouched(event);
        }
    }

    private void performRelease(MotionEvent event) {
        smoother.resetTo(event.getX(), event.getY());
        if (rainbowInteractionListener != null) {
            rainbowInteractionListener.onSketchReleased(event);
        }
    }

    private void performMove(MotionEvent event) {
        smoother.moveTo(event.getX(), event.getY());
        if (rainbowInteractionListener != null) {
            rainbowInteractionListener.onFingerDragged(event);
        }
    }

    private void performMotion(MotionEvent event) {
        if (rainbowInteractionListener != null) {
            rainbowInteractionListener.onMotionEvent(event);
        }
    }

    public boolean isScreenTouched() {
        return screenTouched;
    }

    public boolean isFingerMoving() {
        return fingerMoving;
    }

    /**
     * Used to set a RainbowInteractionListener which will listen for different interaction events
     *
     * @param rainbowInteractionListener
     */
    public void attach(RainbowInteractionListener rainbowInteractionListener) {
        this.rainbowInteractionListener = rainbowInteractionListener;
    }

    public RainbowDrawer getRainbowDrawer() {
        return rainbowDrawer;
    }

    /**
     * Used to remove the attached RainbowInteractionListener
     * Also stops any pending input event processing task.
     */
    public void detach() {
        this.rainbowInteractionListener = null;
        clearEvents();
    }

    public float getX() {
        return x;
    }

    public float getY() {
        return y;
    }

    public float getPreviousX() {
        return px;
    }

    public float getPreviousY() {
        return py;
    }

    public MovementDirection getVerticalDirection() {
        if (smoother.getY() > smoother.getOldY()) {
            return MovementDirection.DOWN;
        } else {
            return MovementDirection.UP;
        }
    }

    public MovementDirection getHorizontalDirection() {
        if (smoother.getX() > smoother.getOldX()) {
            return MovementDirection.RIGHT;
        } else {
            return MovementDirection.LEFT;
        }
    }

    public float getSmoothX() {
        return smoother.getX();
    }

    public float getSmoothY() {
        return smoother.getY();
    }

    public float getPreviousSmoothX() {
        return smoother.getOldX();
    }

    public float getPreviousSmoothY() {
        return smoother.getOldY();
    }

    public float getFingerVelocity() {
        return smoother.getFingerVelocity();
    }

    public enum MovementDirection {
        UP, DOWN, LEFT, RIGHT
    }

    public interface RainbowInteractionListener {
        void onSketchTouched(final MotionEvent event);

        void onSketchReleased(final MotionEvent event);

        void onFingerDragged(final MotionEvent event);

        void onMotionEvent(final MotionEvent event);
    }
}

File: rainbow-lib/src/main/java/com/juankysoriano/rainbow/core/touch/RainbowInputController.kt
package com.juankysoriano.rainbow.core.touch

import android.view.MotionEvent
import com.juankysoriano.rainbow.core.schedulers.RainbowScheduler
import com.juankysoriano.rainbow.core.schedulers.RainbowSchedulers

private const val DIVISIONS = 2

class RainbowInputController
internal constructor(private val scheduler: RainbowScheduler = RainbowSchedulers.single("InputController", RainbowSchedulers.Priority.NORMAL),
                     private val smoother: FingerPositionSmoother = FingerPositionSmoother()) {


    var x: Float = 0f
        private set
    var y: Float = 0f
        private set
    var previousX: Float = 0f
        private set
    var previousY: Float = 0f
        private set
    var isScreenTouched: Boolean = false
        private set
    var isFingerMoving: Boolean = false
        private set

    val verticalDirection: MovementDirection
        get() = if (smoother.y > smoother.oldY) {
            MovementDirection.DOWN
        } else {
            MovementDirection.UP
        }

    val horizontalDirection: MovementDirection
        get() = if (smoother.x > smoother.oldX) {
            MovementDirection.RIGHT
        } else {
            MovementDirection.LEFT
        }

    val smoothX: Float
        get() = smoother.x

    val smoothY: Float
        get() = smoother.y

    val previousSmoothX: Float
        get() = smoother.oldX

    val previousSmoothY: Float
        get() = smoother.oldY

    val fingerVelocity: Float
        get() = smoother.fingerVelocity

    private var rainbowInteractionListener: RainbowInteractionListener? = null

    init {
        previousY = -1f
        previousX = previousY
        y = previousX
        x = y
    }

    fun postEvent(motionEvent: MotionEvent) {
        scheduler.scheduleNow(inputEventFor(MotionEvent.obtain(motionEvent)))
    }

    private fun inputEventFor(motionEvent: MotionEvent): Runnable {
        return Runnable {
            when (motionEvent.action) {
                MotionEvent.ACTION_UP, MotionEvent.ACTION_DOWN -> process(motionEvent)
                MotionEvent.ACTION_MOVE -> splitIntoMultipleEvents(motionEvent)
            }//no-op
        }
    }

    private fun process(motionEvent: MotionEvent) {
        preHandleEvent(motionEvent)
        handleSketchEvent(motionEvent)
        postHandleEvent()
    }

    private fun splitIntoMultipleEvents(event: MotionEvent) {
        val px = x
        val py = y
        val diffX = event.x - px
        val diffY = event.y - py

        for (i in 1..DIVISIONS) {
            val newEventX = px + diffX * i / DIVISIONS
            val newEventY = py + diffY * i / DIVISIONS
            val subEvent = obtainEventWithNewPosition(event, newEventX, newEventY)
            process(subEvent)
        }
    }

    private fun obtainEventWithNewPosition(event: MotionEvent, newEventX: Float, newEventY: Float): MotionEvent {
        val motionEvent = MotionEvent.obtain(event)
        motionEvent.setLocation(newEventX, newEventY)
        return motionEvent
    }

    private fun preHandleEvent(motionEvent: MotionEvent) {
        x = motionEvent.x
        y = motionEvent.y
    }

    private fun handleSketchEvent(event: MotionEvent) {

        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                isScreenTouched = true
                isFingerMoving = false
                performTouch(event)
            }
            MotionEvent.ACTION_UP -> {
                isScreenTouched = false
                isFingerMoving = false
                performRelease(event)
            }
            MotionEvent.ACTION_MOVE -> {
                isScreenTouched = true
                isFingerMoving = true
                performMove(event)
            }
        }

        performMotion(event)
    }

    private fun postHandleEvent() {
        previousX = x
        previousY = y
    }

    private fun performTouch(event: MotionEvent) {
        smoother.resetTo(event.x, event.y)
        rainbowInteractionListener?.onSketchTouched(event)
    }

    private fun performRelease(event: MotionEvent) {
        smoother.resetTo(event.x, event.y)
        rainbowInteractionListener?.onSketchReleased(event)
    }

    private fun performMove(event: MotionEvent) {
        smoother.moveTo(event.x, event.y)
        rainbowInteractionListener?.onFingerDragged(event)
    }

    private fun performMotion(event: MotionEvent) {
        rainbowInteractionListener?.onMotionEvent(event)
    }

    /**
     * Used to set a RainbowInteractionListener which will listen for different interaction events
     *
     * @param rainbowInteractionListener the listener to be set
     */
    fun attach(rainbowInteractionListener: RainbowInteractionListener) {
        this.rainbowInteractionListener = rainbowInteractionListener
    }

    /**
     * Used to remove the attached RainbowInteractionListener
     * Also stops any pending input event processing task.
     */
    fun detach() {
        this.rainbowInteractionListener = null
        this.scheduler.shutdown()
    }

    enum class MovementDirection {
        UP, DOWN, LEFT, RIGHT
    }

    interface RainbowInteractionListener {
        fun onSketchTouched(event: MotionEvent)

        fun onSketchReleased(event: MotionEvent)

        fun onFingerDragged(event: MotionEvent)

        fun onMotionEvent(event: MotionEvent)
    }
}
